### PR TITLE
Merge segments before sending a webhook #742

### LINKED ISSUE #742: Merge segments before sending a webhook
Now we get such a process after detection:
1. Send a webhook
2. Merge segments
3. Insert segments into DB

We should swap 1 and 2

### PR DESCRIPTION
**Pull request by rozetko**
Closes #742 

## Changes
- refactor `insertSegments`
- merge and insert segments before sending a webhook
- update tests

### CODE REVIEW

File: server/src/models/segment_model.ts [hunk 1]
```diff
@@ -93,81 +97,50 @@ export async function findMany(id: AnalyticUnitId, query: FindManyQuery): Promis
   return segs.map(Segment.fromObject);
 }
 
-export async function insertSegments(segments: Segment[]) {
+export async function mergeAndInsertSegments(segments: Segment[]): Promise<SegmentId[]> {
```
Comment (jonyrock) [inline server/src/models/segment_model.ts]: please add a doc comment for the function

File: server/src/models/segment_model.ts [hunk 2]
```diff
@@ -189,3 +162,25 @@ export async function setSegmentsDeleted(ids: SegmentId[]) {
 export function removeSegments(idsToRemove: SegmentId[]) {
   return db.removeMany(idsToRemove);
 }
+
+async function isIntersectedWithLabeled(segment: Segment) {
```
Comment (jonyrock) [inline server/src/models/segment_model.ts]: add return type

File: server/src/controllers/analytics_controller.ts [hunk 3]
```diff
@@ -59,27 +59,29 @@ function onTaskResult(taskResult: TaskResult) {
   }
 }
 
-async function onDetect(detectionResult: DetectionResult) {
+async function onDetect(detectionResult: DetectionResult): Promise<Segment.SegmentId[]> {
```
Comment (jonyrock) [inline server/src/controllers/analytics_controller.ts]: on detect what?

Comment (jonyrock) [reply server/src/controllers/analytics_controller.ts]: please write a comment

File: server/src/controllers/analytics_controller.ts [hunk 4]
```diff
@@ -59,27 +59,42 @@ function onTaskResult(taskResult: TaskResult) {
   }
 }
 
-async function onDetect(detectionResult: DetectionResult) {
+/**
+ * Processes detection result from analytics` DETECT message
+ * Returns IDs of segments inserted into DB if there was no merging
```
Comment (jonyrock) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
 * @returns IDs of segments inserted into DB if there was no merging
```

more about this: https://github.com/Microsoft/tsdoc
https://github.com/Microsoft/tsdoc

File: server/src/models/segment_model.ts [hunk 5]
```diff
@@ -86,88 +90,75 @@ export async function findMany(id: AnalyticUnitId, query: FindManyQuery): Promis
...
   }
   return segs.map(Segment.fromObject);
 }
 
-export async function insertSegments(segments: Segment[]) {
+/**
+ * Merges an array of segments with ones existing in the DB
+ * Inserts resulting segments into DB
+ * @param segments segments to be inserted
+ * Returns IDs of segments inserted into DB
```
Comment (rozetko) [inline server/src/models/segment_model.ts]:
```suggestion
 * @returns IDs of segments inserted into DB
```

**Pull request merged**